tokio/src/runtime/basic_scheduler.rs

```diff
@@ -325,3 +407,45 @@ impl Wake for Shared {
         arc_self.unpark.unpark();
     }
 }
+
+// ===== InnerGuard =====
+
+/// Used to ensure we always place the Inner value
+/// back into its slot in `BasicScheduler`, even if the
+/// future panics.
+struct InnerGuard<'a, P: Park> {
+    inner: Option<Inner<P>>,
+    scheduler: &'a BasicScheduler<P>,
+}
+
+impl<P: Park> InnerGuard<'_, P> {
+    fn block_on<F: Future>(&mut self, future: F) -> F::Output {
+        // The only time inner gets set to `None` is if we have dropped
+        // already so this unwrap is safe.
+        self.inner.as_mut().unwrap().block_on(future)
+    }
+}
+
+impl<P: Park> Drop for InnerGuard<'_, P> {
+    fn drop(&mut self) {
+        if let Some(inner) = self.inner.take() {
+            // Avoid a double panic if we are currently panicking and
+            // the lock may be poisoned.
+            if let Ok(lock) = &mut self.scheduler.inner.lock() {
+                lock.replace(inner);
+            } else {
+                if std::thread::panicking() {
+                    return;
+                } else {
+                    panic!("Inner lock poisoned");
+                }
+            }
```

Comment:
...and again, we should just use the guard returned in the `Err` if the lock is poisoned so we can still put everything back together.